xonsh/history/json.py

```diff
@@ -118,31 +136,46 @@ def run(self):
         if rmfiles_fn is None:
             raise ValueError("Units type {0!r} not understood".format(units))
 
-        for _, _, f in rmfiles_fn(hsize, files):
-            try:
-                os.remove(f)
-            except OSError:
-                pass
+        size_over, rm_files = rmfiles_fn(hsize, files)
+        hist = getattr(builtins.__xonsh__, "history", None)
+        if hist:  # remember last gc pass history size
+            hist.hist_size = size_over + hsize
+            hist.hist_units = units
+
+        if self.force_gc or size_over < hsize:
+            for _, _, f, _ in rm_files:
+                try:
+                    os.remove(f)
+                except OSError:
+                    pass
+        else:
+            print(
+                f"Warning: History garbage collection would discard more history ({size_over} {units}) than it would keep ({hsize}).\n"
+                "Not removing any history for now. Either increase your limit ($XONSH_HIST_SIZE), or run `history gc --force`.",
+                file=sys.stderr,
+            )
 
     def files(self, only_unlocked=False):
         """Find and return the history files. Optionally locked files may be
         excluded.
 
         This is sorted by the last closed time. Returns a list of
-        (timestamp, number of cmds, file name) tuples.
+        (file_size, timestamp, number of cmds, file name) tuples.
         """
-        # pylint: disable=no-member
-        env = getattr(builtins, "__xonsh__.env", None)
+
+        env = getattr(getattr(builtins, "__xonsh__", None), "env", None)
         if env is None:
             return []
+
         boot = uptime.boottime()
         fs = _xhj_get_history_files(sort=False)
         files = []
         for f in fs:
             try:
-                if os.path.getsize(f) == 0:
+                cur_file_size = os.path.getsize(f)
+                if cur_file_size == 0:
                     # collect empty files (for gc)
-                    files.append((time.time(), 0, f))
+                    files.append(cur_file_size, (time.time(), 0, f))
```

Comment:
I'm very confused between this line and the rest of the comments regarding the contents of this `files` array. How does this line work?

Reply:
Thanks for the close review!  First off, this line is a bug.  `cur_file_size` should be 4th element of the tuple (changed design, missed this leftover).

Second, I'm also confused by the special case treatment of zero-length history files.  It may have been a problem once upon a time...  But here, I'm pretty sure the timestamp should be the modification date of the file (which it should have even with zero length), otherwise it might not get garbage collected if user configures 's' age limit.  I'll make both changes and maybe get lucky in CI.